base/stats.go

```diff
@@ -0,0 +1,747 @@
+}
+
+type CBLReplicationPushStats struct {
+	AttachmentPushBytes *SgwIntStat `json:"attachment_push_bytes"`
+	AttachmentPushCount *SgwIntStat `json:"attachment_push_count"`
+	DocPushCount        *SgwIntStat `json:"doc_push_count"`
+	ProposeChangeCount  *SgwIntStat `json:"propose_change_count"`
+	ProposeChangeTime   *SgwIntStat `json:"propose_change_time"`
+	SyncFunctionCount   *SgwIntStat `json:"sync_function_count"`
+	SyncFunctionTime    *SgwIntStat `json:"sync_function_time"`
+	WriteProcessingTime *SgwIntStat `json:"write_processing_time"`
+}
+
+type DatabaseStats struct {
+	AbandonedSeqs           *SgwIntStat `json:"abandoned_seqs"`
+	ConflictWriteCount      *SgwIntStat `json:"conflict_write_count"`
+	Crc32MatchCount         *SgwIntStat `json:"crc32c_match_count"`
+	DCPCachingCount         *SgwIntStat `json:"dcp_caching_count"`
+	DCPCachingTime          *SgwIntStat `json:"dcp_caching_time"`
+	DCPReceivedCount        *SgwIntStat `json:"dcp_received_count"`
+	DCPReceivedTime         *SgwIntStat `json:"dcp_received_time"`
+	DocReadsBytesBlip       *SgwIntStat `json:"doc_reads_bytes_blip"`
+	DocWritesBytes          *SgwIntStat `json:"doc_writes_bytes"`
+	DocWritesBytesBlip      *SgwIntStat `json:"doc_writes_bytes_blip"`
+	DocWritesXattrBytes     *SgwIntStat `json:"doc_writes_xattr_bytes"`
+	HighSeqFeed             *SgwIntStat `json:"high_seq_feed"`
+	NumDocReadsBlip         *SgwIntStat `json:"num_doc_reads_blip"`
+	NumDocReadsRest         *SgwIntStat `json:"num_doc_reads_rest"`
+	NumDocWrites            *SgwIntStat `json:"num_doc_writes"`
+	NumReplicationsActive   *SgwIntStat `json:"num_replications_active"`
+	NumReplicationsTotal    *SgwIntStat `json:"num_replications_total"`
+	NumTombstonesCompacted  *SgwIntStat `json:"num_tombstones_compacted"`
+	SequenceAssignedCount   *SgwIntStat `json:"sequence_assigned_count"`
+	SequenceGetCount        *SgwIntStat `json:"sequence_get_count"`
+	SequenceIncrCount       *SgwIntStat `json:"sequence_incr_count"`
+	SequenceReleasedCount   *SgwIntStat `json:"sequence_released_count"`
+	SequenceReservedCount   *SgwIntStat `json:"sequence_reserved_count"`
+	WarnChannelsPerDocCount *SgwIntStat `json:"warn_channels_per_doc_count"`
+	WarnGrantsPerDocCount   *SgwIntStat `json:"warn_grants_per_doc_count"`
+	WarnXattrSizeCount      *SgwIntStat `json:"warn_xattr_size_count"`
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
+}
+
+type ExpVarMapWrapper struct {
+	*expvar.Map
+}
+
+func (wrapper *ExpVarMapWrapper) MarshalJSON() ([]byte, error) {
+	return []byte(wrapper.String()), nil
+}
+
+type ReplicatorStats struct {
+	checkedSentDesc,
+	numAttachmentBytesTransferred,
+	numAttachmentsTransferred,
+	numDocsFailedToPush,
+	numDocsPushed *prometheus.Desc
```

Comment:
I have switched to having desc for each item. I think this approach works well as they're all in one place and as we have to initialize the map in that method too it makes sense to also initialize the descriptions there. If you think otherwise I can think again and move them to individual vars.

Reply:
There are two good reasons to remove the static descriptions from the struct:
- these descriptions are shared across replications, so don't need to be redefined.  They are also only used in one place, so I don't think having them 'all in one place' really adds that much value.
- removing the descriptions simplifies ReplicatorStats to just `type ReplicatorStats *expvar.Map`, and avoids the need to target ReplicatorStats.topLevelMap everywhere

Reply:
Yeah thats fair. I've made them vars instead and initialize them in the NewSyncGatewayStats() constructor.